tsl/src/dist_util.c

```diff
@@ -113,6 +127,33 @@ dist_util_set_id(Datum dist_id)
 					 (errmsg("database is already a member of a distributed database"))));
 	}
 
+	if (check_metadata)
+	{
+		uuid = ts_metadata_get_value(CStringGetDatum(METADATA_UUID_KEY_NAME),
+									 CSTRINGOID,
+									 UUIDOID,
+									 &isnull);
+		if (isnull)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_DATA_NODE_INVALID_CONFIG),
+					 (errmsg("uuid not set for data node"))));
+		else if (uuid_matches(uuid, dist_id))
+		{
+			Datum dist_uuid_str = DirectFunctionCall1(uuid_out, dist_id);
+
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_DATA_NODE_SELF_REFERENCE),
+					 (errmsg("attempting to add self as data node"),
+					  errdetail("Trying to set dist_uuid to %s, but uuid is the same, so this "
+								"would create a cycle.",
+								DatumGetCString(dist_uuid_str)),
```

Comment:
Too much implementation detail that a user knows little about and too little explanation IMO. I suggest:

```suggestion
					  errdetail("A database that serves as an access node cannot simultaneously be a data node."),
```